httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManager.java

```diff
@@ -113,10 +116,12 @@
     private final HttpClientConnectionOperator connectionOperator;
     private final ManagedConnPool<HttpRoute, ManagedHttpClientConnection> pool;
     private final HttpConnectionFactory<ManagedHttpClientConnection> connFactory;
+    private final ConcurrentMap<HttpRoute, SocketConfig> socketConfigMap;
+    private final ConcurrentMap<HttpRoute, ConnectionMgmtConfig> connectionMgmtConfigMap;
```

Comment:
Should we worry about routes building up over time without a path to removal and garbage collection?

I wonder if we should provide an interface to optionally build a route-specific socket and ConnMgmtConfig the first time a route is used, which can be forgotten and recalculated later if all connections to that route are cleared (or even naively recalculated for each connection, allowing implementations to cache internally).

```java
interface RouteBasedSocketConfigProvider {
    SocketConfig socketConfigForRoute(HttpRoute route);
}
```

Reply:
@carterkozak Using a `Provider` interface for the same end is a great idea. Thank you! 
I will re-work the PR.